_posts/2020-03-02-tomcat-thread-model.md

```diff
@@ -0,0 +1,96 @@
+```json
+{
+  "request": {
+    "mbean": "Catalina:name=\"http-bio-8080\",type=ThreadPool",
+    "attribute": "currentThreadsBusy",
+    "type": "read"
+  },
+  "value": 1,
+  "timestamp": 1564763927,
+  "status": 200
+}
+```
+Warto skonfigurować alerting, który będzie ostrzegał o wysycaniu się puli połączeń i pozwoli szybciej zdiagnozować problem.
+
+## Porównanie connectorów blokujących (bio) i nieblokujących (nio)
+
+### Różnica w działaniu
+
+W Tomcacie 7.x domyślną implementacją connectora jest implementacja blokująca. W Tomcacie 8 i wyższych domyślny jest connector nieblokujący. Podstawowa różnica w działaniu polega na tym, że **w connectorze blokującym wątek przypisywany jest do połączenia, a w connectorze nieblokującym do pojedynczego żądania**. Może to mieć duże znaczenie przy zastosowaniu połączeń keep-alive kiedy to jedno połączenie jest wykorzystywane do przesłania wielu żądań i odpowiedzi. W takim modelu wątek na serwerze jest zajęty na cały czas trwania pojedynczego połączenia mimo, że faktyczne przetwarzanie na serwerze zachodzi tylko w obrębie pojedynczego żądania.
+
+### Test
+
+Dla porównania można wykonać prosty test na dwóch rodzajach connectorów. W teście klient raz na sekundę wysyła żądanie do serwera. Łącznie jeden klient wysyła 30 żądań. Uruchamiamy 20 klientów w ciągu 2 sekund. Jednocześnie monitorujemy rozmiar puli wątków.
+
+Tak wygląda wykres czasów odpowiedzi dla connectora blokującego:
+
+![gatling-bio.png](/assets/img/posts/2020-03-02-tomcat-thread-model/gatling-bio.png)
+
+a tak dla connectora nieblokującego:
+
+![gatling-nio.png](/assets/img/posts/2020-03-02-tomcat-thread-model/gatling-nio.png) 
+
+Nie widać tutaj jakiejś specjalnej różnicy między działaniem connectora bio i nio. Zupełnie inaczej wyglądają natomiast wykresy liczby zajętych wątków w puli. Dla connectora blokującego liczba zajętych wątków podczas trwania testu przekracza 20. Czyli jest zgodna z liczbą klientów wysyłających żądania do serwera. Połączenia keep-alive mają timeout równy 5 sekund więc klient utrzymuje cały czas jedno połączenie do wszystkich żądań:
+
+![threads-bio.png](/assets/img/posts/2020-03-02-tomcat-thread-model/threads-bio.png)
+
+Dla connectora nieblokującego żądania są obsługiwane w większości przez jeden wątek:
+
+![threads-nio.png](/assets/img/posts/2020-03-02-tomcat-thread-model/threads-nio.png)
+
+Test pokazuje, że przy takim modelu przetwarzania żądań za pomocą connectora nieblokującego możemy uzyskać sporą oszczędność zasobów i lepiej zutylizować serwer.
+
+### Użycie mod_proxy zmienia sytuację
+
+Warto zwrócić uwagę na to, że takie wyniki osiągamy w przypadku bezpośredniego połączenia pomiędzy klientem a serwerem Tomcat. Zdarza się jednak, że pomiędzy serwerem Tomcat a klientem mamy jeszcze warstwę pośrednią np. w postaci serwera Apache httpd i modułu mod_proxy. Domyślnie mod_proxy nie ma włączonej opcji keep-alive więc wszystkie połączenia do backendu są zamykane po obsłużeniu pojedynczego żądania. 
```

Comment:
przecinek:
Warto zwrócić uwagę na to, że takie wyniki osiągamy w przypadku bezpośredniego połączenia pomiędzy klientem a serwerem Tomcat.
==>
Warto zwrócić uwagę na to, że takie wyniki osiągamy w przypadku bezpośredniego połączenia pomiędzy klientem, a serwerem Tomcat.

Reply:
Poprawione.